File: app/csharp/AccountingSystem.Infrastructure/MigrationRunner.cs
using FluentMigrator.Runner;
using Microsoft.Extensions.DependencyInjection;

namespace AccountingSystem.Infrastructure
{
    public class MigrationRunner
    {
        public static void MigrateDatabase(string connectionString, string databaseType = "PostgreSQL")
        {
            var serviceProvider = CreateServices(connectionString, databaseType);

            using (var scope = serviceProvider.CreateScope())
            {
                UpdateDatabase(scope.ServiceProvider);
            }
        }

        private static IServiceProvider CreateServices(string connectionString, string databaseType)
        {
            var services = new ServiceCollection();

            if (databaseType == "PostgreSQL")
            {
                services.AddFluentMigratorCore()
                    .ConfigureRunner(rb => rb
                        .AddPostgres()
                        .WithGlobalConnectionString(connectionString)
                        .ScanIn(typeof(MigrationRunner).Assembly).For.Migrations())
                    .AddLogging(lb => lb.AddFluentMigratorConsole());
            }
            else if (databaseType == "MySQL")
            {
                services.AddFluentMigratorCore()
                    .ConfigureRunner(rb => rb
                        .AddMySql5()
                        .WithGlobalConnectionString(connectionString)
                        .ScanIn(typeof(MigrationRunner).Assembly).For.Migrations())
                    .AddLogging(lb => lb.AddFluentMigratorConsole());
            }

            return services.BuildServiceProvider(false);
        }

        private static void UpdateDatabase(IServiceProvider serviceProvider)
        {
            var runner = serviceProvider.GetRequiredService<IMigrationRunner>();
            runner.MigrateUp();
        }
    }
}

File: app/fsharp/AccountingSystem.Infrastructure/MigrationRunner.fs
module AccountingSystem.Infrastructure.MigrationRunner

open FluentMigrator.Runner
open Microsoft.Extensions.DependencyInjection
open System

let migrateDatabase (connectionString: string) (databaseType: string) =
    let serviceProvider =
        let services = ServiceCollection()

        let configureRunner (rb: IMigrationRunnerBuilder) =
            match databaseType with
            | "PostgreSQL" ->
                rb.AddPostgres()
                  .WithGlobalConnectionString(connectionString)
                  .ScanIn(typeof<Migrations.Migration_20250106_001_InitialSetup>.Assembly)
                  .For.Migrations() |> ignore
            | "MySQL" ->
                rb.AddMySql5()
                  .WithGlobalConnectionString(connectionString)
                  .ScanIn(typeof<Migrations.Migration_20250106_001_InitialSetup>.Assembly)
                  .For.Migrations() |> ignore
            | _ ->
                failwith $"Unsupported database type: {databaseType}"

        services.AddFluentMigratorCore()
                .ConfigureRunner(Action<IMigrationRunnerBuilder> configureRunner)
                .AddLogging(fun lb -> lb.AddFluentMigratorConsole() |> ignore)
                .BuildServiceProvider(false)

    use scope = serviceProvider.CreateScope()
    let runner = scope.ServiceProvider.GetRequiredService<IMigrationRunner>()
    runner.MigrateUp()
